### ISSUE TITLE
cleanup and fix getrusage03

### ISSUE THREAD

**Issue by metan-ucw** [linux-test-project/ltp#802]
The test should be rewritten to the new library but apart from that there are couple of other problems as well. Namely:

* The last test exec_without_fork() exits the test process which means that the test results are not propagated outside

* The test uses sleep(1) in order to wait for child processes, looks like we should use tst_process_state_wait() instead

* The DELTA_MAX is probably too small for machines with 64k pages, e.g. ppc64le

**Comment by ZiyaoXie** [comment]
Hi, Cyril,

I'm trying to rewrite this testsuite. For the first point, is the following code an appropriate solution?

```C
static void exec_without_fork(void)
{
    if ((pid = SAFE_FORK()) == 0) {
        char str_maxrss_self[BUFSIZ], str_maxrss_child[BUFSIZ];

        SAFE_GETRUSAGE(RUSAGE_SELF, &ru);
        sprintf(str_maxrss_self, "%ld", ru.ru_maxrss);
        SAFE_GETRUSAGE(RUSAGE_CHILDREN, &ru);
        sprintf(str_maxrss_child, "%ld", ru.ru_maxrss);

        SAFE_EXECLP("getrusage03_child", "getrusage03_child", "-v", "-s", str_maxrss_self,
                    "-l", str_maxrss_child, NULL);
    }

    SAFE_WAITPID(pid, &status, 0);
    check_return(WIFEXITED(status) && WEXITSTATUS(status) == 0,
                 "exec without fork succeeded", "exec without fork failed");
}
```

Kind regards,
Ziyao

**Comment by metan-ucw** [comment]
I guess that we can use `tst_reap_children()` instead of the `SAFE_WAITPID()` and `check_return()` but other than that the change looks fine.

**Comment by ZiyaoXie** [comment]
Hi @mdoucha，thanks for your reminder, I will pay attention to add swapping check during the rewrite.

In addition, I also noticed `exec_without_fork()`, which I prefer to call `inherit_exec()`.

I guess that the function is written to test `initial ~= exec`, since I found the line in the `getrusage` manual page:

```Text
Resource usage metrics are preserved across an execve(2).
```

IMHO, writing this way may not change the test logic, so I'll try to modify only the function name but keep `SAFE_FORK()`, thx.

**Comment by ZiyaoXie** [comment]
Hi @metan-ucw @mdoucha , any suggestions on the value of `DELTA_MAX`? 

As for checking the `VmSwap`, I intend to add it to `consum()`, like:
```C
static void consume(int consume_nr)
{
    void* ptr;
    size_t size;
    unsigned long vmswap_size;

    size = consume_nr * 1024 * 1024;
    ptr = SAFE_MALLOC(size);
    memset(ptr, 0, size);

    SAFE_FILE_LINES_SCANF("/proc/self/status", "VmSwap: %lu", &vmswap_size);
    if (vmswap_size > 0)
        tst_brk(TBROK, "VmSwap is not zero");
    tst_res(TINFO, "allocate %dMB", consume_nr);
}
```

Thx.

**Comment by mdoucha** [comment]
> Hi @metan-ucw @mdoucha , any suggestions on the value of `DELTA_MAX`?

I'd say 256 pages should be enough. `DELTA_MAX` only needs to account for some stack overhead. But if the page size gets really huge on some future platform, the subtests may need larger gaps than 100MB to avoid overlaps. I'm fine with either increasing the consumed amount by `2*DELTA_MAX` between subtests or just throwing `TBROK` when the pagesize is too big.

> As for checking the `VmSwap`, I intend to add it to `consum()`, like:

Looks good.

**Comment by ZiyaoXie** [comment]
Hi @metan-ucw @mdoucha , thanks for your suggestions, I made the following changes:

- Rewrite `exec_without_fork()` to `inherit_exec()` and make the results can be reported in the test process.
- Replace `sleep(1)` with `TST_PROCESS_STATE_WAIT()` and `TST_PROCESS_RELEASE_WAIT()`, which returns a value indicating if the process is released with checking whether `/proc/pid` exists.
- Double the value of `DELTA_MAX`, but not sure about that.
- Add swapping check to `consum()`.

In addition, I made other modifications:

- Remove `fork_malloc()`, since maybe this point can be covered by `inherit_fork1()` and `inherit_fork2()`.

More details in [patches](https://patchwork.ozlabs.org/project/ltp/list/?series=249160), thx.
